src/util/interval.h

```diff
@@ -0,0 +1,472 @@
+/*******************************************************************\
+
+ Module: intervals
+
+ Author: Daniel Neville (2017)
+
+\*******************************************************************/
+#ifndef CPROVER_UTIL_INTERVAL_H
+#define CPROVER_UTIL_INTERVAL_H
+
+#include <util/arith_tools.h>
+#include <util/expr.h>
+#include <util/namespace.h>
+#include <util/simplify_expr.h>
+#include <util/std_expr.h>
+#include <util/std_types.h>
+#include <util/symbol_table.h>
+#include <util/threeval.h>
+
+#include <iostream>
+#include <ostream>
+#include <sstream>
+
+/// +∞ upper bound for intervals
+class max_exprt : public exprt
+{
+public:
+  explicit max_exprt(const typet &_type) : exprt(ID_max, _type)
+  {
+  }
+
+  explicit max_exprt(const exprt &_expr) : exprt(ID_max, _expr.type())
+  {
+  }
+};
+
+/// -∞ upper bound for intervals
+class min_exprt : public exprt
+{
+public:
+  explicit min_exprt(const typet &_type) : exprt(ID_min, _type)
+  {
+  }
+
+  explicit min_exprt(const exprt &_expr) : exprt(ID_min, _expr.type())
+  {
+  }
+};
+
+/// Represents an interval of values.
+/// Bounds should be constant expressions
+/// or min_exprt for the lower bound
+/// or max_exprt for the upper bound
+/// Also, lower bound should always be <= upper bound
+class constant_interval_exprt : public binary_exprt
```

Comment:
:pick: from an "is-a" relation, I don't think an interval really is a binary expr? Not sure...